spring-data-microservice-transaction-sample/customer-service/src/main/java/sample/customer/domain/repository/CustomerRepository.java

```diff
@@ -0,0 +1,31 @@
+package sample.customer.domain.repository;
+
+import com.scalar.db.sql.springdata.twopc.ScalarDbTwoPcRepository;
+import java.util.function.Supplier;
+import org.springframework.stereotype.Repository;
+import org.springframework.transaction.annotation.Transactional;
+import sample.customer.domain.model.Customer;
+
+@Transactional
+@Repository
+public interface CustomerRepository extends ScalarDbTwoPcRepository<Customer, Integer> {
+
+  default void insertIfNotExists(Customer customer) {
+    if (!findById(customer.customerId).isPresent()) {
+      insert(customer);
+    }
+  }
+
+  default <T> T execOneshotOperation(Supplier<T> task) {
```

Comment:
These API should be added to Spring Data JDBC for ScalarDB. I'll handle it later

Reply:
Or can we make a repository class inherit both `ScalarDbRepository` and `ScalarDbTwoPcRepository` to execute both normal transactions and 2PC transactions?

Reply:
Good point! It's a bit complicated story, but let me explain it.

`ScalarDbTwoPcRepository` inherits `ScalarDbRepository` and it can already execute CRUD (and Scan) operations as well as `ScalarDbRepository`. 

Because of the separated modules in Spring-Tx and Spring Data, the following capabilities are handled by different components and configurations in Spring Data JDBC for ScalarDB:

1. avoid automatic implicit commit when quitting a method of ScalarDbTwoPcRepository
   - handled by `ScalarDbSuspendableTransactionManager`
2. provide APIs to call the microservice transaction APIs of ScalarDB
   - handled by `ScalarDbTwoPcRepository`
3. use ScalarDB `two_phase_commit_transaction` mode
   - configured as a part of DataSource in [application.properties](https://github.com/scalar-labs/scalardb-samples/pull/37/files#diff-ea7b1f4bd859b6b61f5164af64fe8ab18a9fc8241dd013dbd0ec1f7c5d2d87a8R16)

As for 3, Spring Data basically supports only a single DataSource and this sample application also uses a single DataSource with 2PC transaction mode. I've already established how to use multiple DataSources (https://github.com/scalar-labs/scalardb-sql/blob/main/spring-data/example/mixed-transaction-modes/src/main/resources/application.properties), in which we use both normal and 2PC transaction modes in a single application, but it requires users to implement a bit complicated Spring Configuration classes. So I decided to use a single DataSource with 2PC mode. It also results in `execOneshotOperation()` is required.

After adding `execOneshotOperation()` in Spring Data JDBC for ScalarDB, we can remove the method from this sample application.

So, regarding the questions,

> This means ScalarDbTwoPcRepository can execute both normal transactions and 2PC transactions, correct?

Yes. But it also needs to call `begin()`, `commit()` and so on.

> When I looked at this code, I was wondering if we really need a 2PC transaction here because it's a one shot operation. But if this execOneshotOperation() is automatically generated, I'm okay with it.

All Spring Data Repository class inheriting `ScalarDbTwoPcRepository` work in 2PC transaction mode. So `execOneshotOperation()` is needed here for now until we add the same one to Spring Data JDBC for ScalarDB.